File: competitive_programming/leetcode/07_typescript/69_sqrt_x.ts
function sqrtX(x: number): number {
  let low = 0;
  let high = x;
  while (low <= high) {
    const middle = Math.floor((low + high) / 2);
    if (middle * middle <= x) low = middle + 1;
    else high = middle - 1;
  }
  return high;
}

if (sqrtX(4) !== 2 || sqrtX(8) !== 2 || sqrtX(0) !== 0) throw new Error("sqrt x failed");
console.log("69 sqrt x ok");

File: competitive_programming/leetcode/11_golang/69_sqrt_x.go
package main

import "fmt"

func sqrtX(x int) int {
	low, high := 0, x
	for low <= high {
		middle := (low + high) / 2
		if middle*middle <= x {
			low = middle + 1
		} else {
			high = middle - 1
		}
	}
	return high
}

func main() {
	if sqrtX(4) != 2 || sqrtX(8) != 2 || sqrtX(0) != 0 {
		panic("sqrt x failed")
	}
	fmt.Println("69 sqrt x ok")
}
